cos_alerter/daemon.py

```diff
@@ -4,53 +4,119 @@
 
 """Daemon process for COS Alerter."""
 
+import argparse
+import logging
 import signal
-import subprocess
 import sys
+import threading
 import time
+from typing import List, Optional
 
-from .alerter import AlerterState, send_test_notification
+import waitress
+
+from .alerter import AlerterState, config, send_test_notification
+from .server import app
+
+logger = logging.getLogger("cos_alerter.daemon")
+
+LOG_LEVEL_CHOICES = {
+    "CRITICAL": logging.CRITICAL,
+    "critical": logging.CRITICAL,
+    "ERROR": logging.ERROR,
+    "error": logging.ERROR,
+    "WARNING": logging.WARNING,
+    "warning": logging.WARNING,
+    "INFO": logging.INFO,
+    "info": logging.INFO,
+    "DEBUG": logging.DEBUG,
+    "debug": logging.DEBUG,
+}
 
 
 def sigint(_, __):  # pragma: no cover
     """Signal handler to exit cleanly on SIGINT."""
+    logger.info("Received SIGINT.")
+    logger.debug("Exiting.")
     sys.exit()
 
 
 def sigusr1(_, __):  # pragma: no cover
     """Signal handler for SIGUSR1 which sends a test notification."""
-    send_test_notification()
+    logger.info("Received SIGUSR1.")
+    send_thread = threading.Thread(target=send_test_notification)
+    send_thread.start()
+
 
+def parse_args(args: List[str]) -> argparse.Namespace:
+    """Parse the command line arguments."""
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--log-level",
+        choices=list(LOG_LEVEL_CHOICES),
+        help='Logging level. Overrides config value "log_level"',
+    )
+    return parser.parse_args(args=args)
 
-def main(run_for=None):
+
+def main(run_for: Optional[int] = None, argv: List[str] = sys.argv):
     """Main method for COS Alerter.
 
     Args:
         run_for: This argument is for testing purposes. If set, only run for "run_for" seconds.
+        argv: Can be used to override the cli args for testing purposes.
     """
+    args = parse_args(argv[1:])
+
+    if args.log_level:
+        log_level = LOG_LEVEL_CHOICES[args.log_level]
+    else:
+        log_level = LOG_LEVEL_CHOICES[config["log_level"]]
+    cos_logger = logging.getLogger("cos_alerter")
+    waitress_logger = logging.getLogger("waitress")
+    cos_logger.propagate = False
+    waitress_logger.propagate = False
+    cos_logger.setLevel(level=log_level)
+    waitress_logger.setLevel(level=log_level)
+    handler = logging.StreamHandler()
+    formatter = logging.Formatter("%(asctime)s:%(levelname)s:%(name)s:%(message)s")
+    handler.setFormatter(formatter)
+    cos_logger.addHandler(handler)
+    waitress_logger.addHandler(handler)
```

Comment:
I would be nice to have this as a separate block. All of the logging stuff is global, so if it were handled elsewhere (even by importing a module which handled the setup at the top level), the logging would be usable without necessarily going through this entrypoint.

Reply:
Done.